### DISCUSSION TITLE
Replacing lines with issues instead of replacing tokens

### DISCUSSION THREAD

**Discussion by oscarssanchez** [squizlabs/PHP_CodeSniffer#3735] [Q&A] [answered]
Hello! 

I have dug a bit into the code and would like to ask to have a better idea of what would it take for me to implement this:

* When I see PHPCS output, it gives you the line with issues.
* I'd like to replace a whole line that is flagged by the sniffer 
* I see the current method to handle fixes is to replace tokens with issues and recreate the file later, but no ability to replace lines (please correct me if i'm wrong)

I've checked that CBF implements Report, so am wondering if I need to implement that interface and create a new class that bases on CBF (or perhaps, extend CBF for my purposes), or if there's an easier way to achieve this.

Thank you!

**Comment by jrfnl** [comment, marked as answer]
Code can be written in so many different ways that to replace lines is rarely the correct solution and I'd strongly advise against it.
Think:
```php
$foo = function_call($a, $b);
// versus (functionally the same):
$foo = function_call(
  param_b : $b,
  param_a : $a,
);
```

If you really need to replace a whole "line", use the token stream to walk to the start of the line - `$tokens[$ptr]['column'] === 1` - and replace all the tokens until you reach the end of the line - `$tokens[$ptr]['line'] !== $tokens[($ptr + 1)]['line']` (or rather, replace the first token with your "replacement" and then replace the rest of the tokens with `''`, which will effectively remove them).

As you would be touching multiple tokens, make sure to wrap the changes you are making in a changeset to prevent problems if your fix would conflict with other fixes being made.

> I've checked that CBF implements Report, so am wondering if I need to implement that interface and create a new class that bases on CBF (or perhaps, extend CBF for my purposes)

The `Report` interface is only for the handling of the output to the screen, it doesn't have anything to do with the actual fixing.

**Comment by oscarssanchez** [comment]
Thank you for taking the time to reply [rfnl](https://github.com/jrfnl)! I am playing with fixing more complicated linting issues with AI. As you have outlined, replacing code has its risks, but I've found it can resolve specific problems pretty good on a consistent basis (as an example, escaping issues with WordPress coding standards), while others not so well.
